source/base/incremental_function.cc

```diff
@@ -0,0 +1,101 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2007 - 2018 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/incremental_function.h>
+
+#include <deal.II/lac/vector.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Functions
+{
+  template <int dim, typename RangeNumberType>
+  IncrementalFunction<dim, RangeNumberType>::IncrementalFunction(
+    Function<dim, RangeNumberType> &base)
+    : base(base)
+    , delta_t(time_type())
+    , values_old(base.n_components)
+  {}
+
+
+
+  template <int dim, typename RangeNumberType>
+  void
+  IncrementalFunction<dim, RangeNumberType>::set_decrement(
+    const time_type delta)
+  {
+    Assert(delta >= 0.0,
+           ExcMessage("The decrement must be set to a non-negative value."));
+    delta_t = delta;
+  }
+
+
+
+  template <int dim, typename RangeNumberType>
+  RangeNumberType
+  IncrementalFunction<dim, RangeNumberType>::value(const Point<dim> &p,
+                                                   unsigned int      comp) const
+  {
+    // Cache the time state of the base class in case it has been changed
+    // within the user code. We reset the wrapped function to the original
+    // state once we're done with our own evaluations.
+    const auto orig_time = base.get_time();
+
+    base.set_time(this->get_time());
```

Comment:
how does this compile? You're in a `const` function, and you're calling a non-`const` function of a member variable that is not marked as `mutable`. I'm confused.

Reply:
SFINAE and the fact that only `vector_value` is tested I suppose?

Reply:
> how does this compile?

probably because it was `unsigned int` instead of `const unsigned int`, so I think it never compiled.

Reply:
Hm, @tamiko, I don't think I get it. The code here is basically this:
```
  class X { public: void f(); };   // class with a non-const member function

  class Y {
      X &x;                               // non-const reference
    public:
      void foo () const {
         x.f();          // call non-const function 'f' of a reference object from a const function
      }
  };
```

I would have expected this to not compile, but it actually does. I guess I misunderstood that in the `const` member function, the *reference* is `const` but the object pointed to is not.

Reply:
In other words, forget about it.

Reply:
> I guess I misunderstood that in the const member function, the reference is const but the object pointed to is not.

i guess I don't need to make the reference `mutable` then, right?

Reply:
@bangerth Indeed, the corresponding const type is `X & const` (which is simply `X &`, references are always const), and not `const X &`.

Reply:
Yes, the reference is immutable, but the referenced object remains mutable. What pointless semantics :-)

Reply:
@davydden -- correct, no need to make the reference variable `mutable`. Though it probably wouldn't hurt and make the code easier to read!

Reply:
> Though it probably wouldn't hurt and make the code easier to read!

i think that lead to some compiler errors, but I did not bother looking closer.

Reply:
```
error: reference `base` cannot be declared `mutable` [-fpermissive]
```

Reply:
but let's "Resolve conversation" as there's not much we can do here, i think.

Reply:
Correct. Thanks for trying!